src/main/java/spoon/reflect/code/CtTry.java

```diff
@@ -47,6 +47,19 @@ public interface CtTry extends CtStatement, TemplateParameter<Void>, CtBodyHolde
 	@PropertySetter(role = CATCH)
 	<T extends CtTry> T addCatcher(CtCatch catcher);
 
+	/**
+	 * Adds a catch block at the specified position in the <code>try</code> statement.
+	 * Behaves similarly to {@link java.util.List#add(int, Object)}.
+	 *
+	 * @param <T> the type of the try statement
+	 * @param position the position at which the <code>catcher</code> is to be inserted
+	 * @param catcher the catch statement to be inserted
+	 * @return this try statement
+	 * @throws IndexOutOfBoundsException if the position is out of range (position < 0 || position > number of catchers)
+	 */
+	@PropertySetter(role = CATCH)
+	<T extends CtTry> T addCatcherAt(int position, CtCatch catcher);
```

Comment:
That is indeed confusing, but why does it throw the class cast exception? `CtTryWithResource` extends from `CtTry` so it should be valid during runtime as well. Is it because type erasure replaces `<T extends CtTry>` to `CtTry`? I do not understand what exactly is happening at runtime.

Reply:
> CtTryWithResource extends from CtTry so it should be valid during runtime as well.

Yes, but the actual dynamic type of the `tryStatement` variable is `CtTryImpl`, which does not implement `CtTryWithResource`. So, there's an exception when the runtime tries to cast to the latter.

I wrote a really elaborate explanation for this, and then the page refreshed and I lost it :D. I'll give it a go again tomorrow, but the very short explanation is that the problems occur with the unchecked cast in the return statement (i.e. `(T)`). Basically, it's _not_ a cast at all, it's just telling the compiler "trust me mate, this is fine".

The actual cast here occurs when we assign to the variable with type `CtTryWithResource` after the method has returned a value. That's where the crash occurs. Java is hailed as a statically typed language, but it's actually very much a dynamically typed language as well, and implicit type checks occur all the time wherever generics are involved.

I'll try to circle back around to the rather nasty ramifications this stuff can have: a confusing `ClassCastException` is just the tip of the iceberg.

I spent way too much time on writing (and losing) the explanation here so I'll have to get back to the actual PR at a later time :)